vpr/src/route/router_lookahead_sampling.cpp

```diff
@@ -0,0 +1,229 @@
+#include "router_lookahead_sampling.h"
+
+#include <vector>
+
+#include "globals.h"
+#include "vtr_math.h"
+#include "vtr_geometry.h"
+#include "vtr_time.h"
+
```

Comment:
I think we need a file-scope comment here which outlines at a high-level how this sampling method works, as it is currently not clear.

For instance, I get that there is some kind of spatial partitioning and statistics collection about number of segment instances going on, which guides the selection of sample points. But it is not clear how these different ideas fit together, or what the motivation/intuition is behind this approach.

Reply:
Since I wrote this, here's a high level overview:

For each segment type:
1. Compute bounding box for that segment type. Within this bounding box:
2. Count the number of nodes at each location.
3. For each of NxN subsections:
   a. Compute a histogram of counts for each location.
   b. Choose a specific point within a quantile, e.g. 50-70%. This is to select a representative location while avoiding extremities.
4. Sort the points. I'm using a Z-curve for spatial locality, see the comments for details.
5. Find all the nodes that fall within each sample point of the given type.

Maybe the inline comments should be expanded, but the code in `find_sample_regions()` seems pretty readable, although reading the datatypes in the header first might help.